packages/core/src/electron-main/electron-main-application.ts

```diff
@@ -731,6 +710,60 @@ export class ElectronMainApplication {
         }
     }
 
+    protected onWebContentsCreated(event: ElectronEvent, webContents: WebContents): void {
+        // Block any in-page navigation except loading the secondary window contents
+        webContents.on('will-navigate', evt => {
+            if (new URI(evt.url).path.fsPath() !== new Path(this.globals.THEIA_SECONDARY_WINDOW_HTML_PATH).fsPath()) {
+                evt.preventDefault();
+            }
+        });
+
+        webContents.setWindowOpenHandler(details => {
+            // if it's a secondary window, allow it to open
+            if (new URI(details.url).path.fsPath() === new Path(this.globals.THEIA_SECONDARY_WINDOW_HTML_PATH).fsPath()) {
+                const { minWidth, minHeight } = this.getDefaultOptions();
+                const options: BrowserWindowConstructorOptions = {
+                    ...this.getDefaultTheiaSecondaryWindowBounds(),
+                    // We always need the native window frame for now because the secondary window does not have Theia's title bar by default.
+                    // In 'custom' title bar mode this would leave the window without any window controls (close, min, max)
+                    // TODO set to this.useNativeWindowFrame when secondary windows support a custom title bar.
+                    frame: true,
+                    minWidth,
+                    minHeight
+                };
+                if (!this.useNativeWindowFrame) {
+                    // If the main window does not have a native window frame, do not show  an icon in the secondary window's native title bar.
+                    // The data url is a 1x1 transparent png
+                    options.icon = nativeImage.createFromDataURL(
+                        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQI12P4DwQACfsD/WMmxY8AAAAASUVORK5CYII=');
+                }
+                return {
+                    action: 'allow',
+                    overrideBrowserWindowOptions: options,
+                };
+            } else {
+                const uri: URI = new URI(details.url);
+                let okToOpen = uri.scheme === 'https' || uri.scheme === 'http';
+                if (!okToOpen) {
+                    const button = dialog.showMessageBoxSync(BrowserWindow.fromWebContents(webContents)!, {
+                        message: nls.localize('theia/core/openLink.message', 'Open link\n\n{0}\n\nin the system handler?', details.url),
```

Comment:
Note: Only the frontend actually loads the localizations. The Electron main app isn't really capable of translating anything (yet). I should maybe look into how we could accomplish this. In the meantime, everything will be displayed in English.

Reply:
Hmh...that sucks. An alternative would be to delegate showing the dialog to the front end, but that would require some back-and-forth again. Hmh....

Reply:
It's fine for me, I just wanted to mention it, since I was a bit confused by the import of `nls` in the electron main thread.

Reply:
Come to think of it: why is `nls` in the `common` folder if it's only available on the browser side?

Reply:
See https://github.com/eclipse-theia/theia/pull/10153#issuecomment-927616277 :)

Reply:
It's mainly due to stuff like this:

https://github.com/eclipse-theia/theia/blob/d79b7101adef9db71e74c13744a3fde4f79a343c/packages/core/src/common/command.ts#L62-L80

The code is in `common` (and is actually only called from `browser`) but requires the nls localization for it to work. We should refactor this sometime.